File: frontend/src/types/scan.ts
export interface ScanBook {
  rank: number
  title: string
  author: string
  genre: string
  words: number
  followers: number
  growth_7d: number
  rating: number
  tags: string[]
}

export interface GenreStat {
  genre: string
  count: number
  avg_followers: number
  avg_growth: number
}

export interface HotTag {
  tag: string
  count: number
}

export interface TrendStats {
  total: number
  insights: string
  genre_distribution: GenreStat[]
  hot_tags: HotTag[]
  top_books: string[]
}

export interface TopicDecision {
  topic: string
  genre: string
  hot_tag: string
  rationale: string
  hooks: string[]
  risk: string
}

export interface ScanSnapshot {
  id: number
  platform: string
  snapshot_at: string | null
  raw: { platform: string; books: ScanBook[] } | null
  cleaned: {
    platform: string
    books: ScanBook[]
    invalid: ScanBook[]
    dropped: number
    stats: TrendStats
    topic_decision: TopicDecision
  } | null
  report: string | null
}

export interface ScanLatest {
  platforms: ScanSnapshot[]
}

File: backend/app/schemas/scan.py
"""扫榜链路输出契约（D9：US-24，§5.1 契约清单）。"""

from pydantic import BaseModel, Field


class ScanBook(BaseModel):
    rank: int = Field(..., ge=1)
    title: str = Field(..., min_length=1)
    author: str = Field(..., min_length=1)
    genre: str = Field(..., min_length=1)
    words: int = Field(..., ge=0)  # 字数（万字）
    followers: int = Field(..., ge=0)  # 收藏 / 在读
    growth_7d: int = Field(..., ge=0)  # 7 日新增收藏
    rating: float = Field(default=0, ge=0, le=10)
    tags: list[str] = Field(default_factory=list)


class ScanRankings(BaseModel):
    platform: str = Field(..., min_length=1)
    books: list[ScanBook] = Field(..., min_length=1)


class TrendReport(BaseModel):
    total: int = Field(..., ge=0)
    insights: str = Field(..., min_length=1)
    genre_distribution: list[dict] = Field(default_factory=list)
    hot_tags: list[dict] = Field(default_factory=list)
    top_books: list[str] = Field(default_factory=list)


class TopicDecision(BaseModel):
    topic: str = Field(..., min_length=1)
    genre: str = Field(..., min_length=1)
    rationale: str = Field(..., min_length=1)
    hooks: list[str] = Field(default_factory=list)
    risk: str = Field(default="", max_length=256)
    hot_tag: str = Field(default="", max_length=32)


class ScanReportOut(BaseModel):
    report: str = Field(..., min_length=1)
